File: source/docker-testnet/typescript-sdk/test/local-build.test.ts
import { describe, test, expect, beforeAll, afterAll } from "bun:test";
import { DockerTestnet } from "../src/index";
import { AptosClient } from "aptos";

/**
 * Local Image Build Integration Test
 *
 * This test suite validates the entire local build workflow:
 * 1. Building the validator image from source
 * 2. Starting a testnet with the locally built image
 * 3. Verifying the testnet is fully functional
 *
 * NOTE: This test is SLOW on first run (~10-15min to build image)
 *       Subsequent runs are faster (~2-3min) thanks to sccache
 *
 * To run:
 *   bun test test/local-build.test.ts
 *
 * To skip (if you don't want to wait):
 *   This test is marked with a timeout and can be interrupted
 */

// Global reference for cleanup
let globalTestnet: DockerTestnet | undefined;
let cleanupInProgress = false;

async function performCleanup(reason: string): Promise<void> {
    if (cleanupInProgress) return;
    cleanupInProgress = true;
    console.log(`\n🛑 ${reason}`);
    if (globalTestnet) {
        try {
            await globalTestnet.teardown();
        } catch (error) {
            console.error("✗ Failed to tear down testnet:", error);
        }
    }
    cleanupInProgress = false;
}

process.on("SIGINT", async () => {
    await performCleanup("Received SIGINT");
    process.exit(130);
});

describe("Local Image Build - Integration Test", () => {
    let testnet: DockerTestnet | undefined;
    let client: AptosClient;
    const NUM_VALIDATORS = 2; // Use 2 validators to reduce resource usage

    /**
     * Step 1: Build the local image
     *
     * This is the slow part on first run. Progress will be shown in the console.
     * sccache will speed up subsequent builds significantly.
     */
    test("should build local image successfully", async () => {
        console.log("\n=== Building Local Validator Image ===");
        console.log("This may take 10-15 minutes on first build...");
        console.log("Subsequent builds will be much faster (~2-3min) thanks to sccache\n");

        await DockerTestnet.buildLocalImage({
            profile: "release",
            showStats: false, // Set to true if you want sccache stats
        });

        console.log("✓ Image built successfully");
    }, 1200000); // 20 minute timeout for first build

    /**
     * Step 2: Start testnet with local image
     *
     * This verifies that:
     * - The docker-compose.yaml correctly uses USE_LOCAL_IMAGE
     * - The locally built image is available
     * - All validators start successfully
     */
    test("should start testnet with local image", async () => {
        console.log("\n=== Starting Testnet with Local Image ===");

        testnet = await DockerTestnet.new(NUM_VALIDATORS, {
            useLocalImage: true,
        });
        globalTestnet = testnet;

        expect(testnet).toBeDefined();
        expect(testnet.getNumValidators()).toBe(NUM_VALIDATORS);

        console.log(`✓ Testnet started with ${NUM_VALIDATORS} validators`);
        console.log(`  Using image: atomica-validator:local`);
    }, 300000); // 5 minute timeout

    /**
     * Step 3: Verify testnet functionality
     *
     * This ensures the locally built image is actually working correctly:
     * - REST API is accessible
     * - Consensus is running (blocks are being produced)
     * - On-chain state is accessible
     */
    test("should have functional validators", async () => {
        expect(testnet).toBeDefined();
        client = new AptosClient(testnet!.validatorApiUrl(0));

        console.log("\n=== Verifying Testnet Functionality ===");

        // Wait for network to stabilize and produce blocks
        console.log("Waiting for consensus to start...");
        await new Promise(resolve => setTimeout(resolve, 5000));

        try {
            await testnet!.waitForBlocks(1, 60);
            console.log("✓ Consensus is running - blocks being produced");
        } catch (e) {
            console.warn("Warning: Timeout waiting for blocks, but continuing...");
        }

        // Query each validator
        for (let i = 0; i < NUM_VALIDATORS; i++) {
            const info = await testnet!.getLedgerInfo(i);

            expect(info.chain_id).toBe(4);
            expect(info.node_role).toBe("validator");
            expect(parseInt(info.ledger_version)).toBeGreaterThan(0);

            console.log(`✓ Validator ${i}: chain_id=${info.chain_id}, version=${info.ledger_version}, height=${info.block_height}`);
        }

        // Verify on-chain data is accessible
        const validatorSet = await client.getAccountResource("0x1", "0x1::stake::ValidatorSet");
        const activeValidators = (validatorSet.data as any).active_validators;

        expect(activeValidators.length).toBeGreaterThanOrEqual(NUM_VALIDATORS);
        console.log(`✓ On-chain state accessible: ${activeValidators.length} active validators`);

        console.log("\n✅ All tests passed - local image is fully functional!");
    }, 120000); // 2 minute timeout

    /**
     * Step 4: Verify git hash metadata
     *
     * The local build should include the git SHA from atomica-aptos source
     */
    test("should have correct build metadata", async () => {
        expect(testnet).toBeDefined();

        const info = await testnet!.getLedgerInfo(0);

        if (info.git_hash) {
            console.log(`✓ Build metadata present: git_hash=${info.git_hash}`);
            expect(info.git_hash).toBeTruthy();
            expect(info.git_hash.length).toBeGreaterThan(6); // Should be a commit hash
        } else {
            console.log("Note: git_hash not present in ledger info");
        }
    }, 30000);

    /**
     * Cleanup after all tests
     */
    afterAll(async () => {
        await performCleanup("Tests completed");
        globalTestnet = undefined;
    });
});

/**
 * Optional: Test sccache effectiveness
 *
 * This test can be run manually to verify sccache is working.
 * It requires running the build twice.
 */
describe("Local Build - sccache Performance (Optional)", () => {
    test.skip("should demonstrate sccache speedup", async () => {
        console.log("\n=== Testing sccache Performance ===");
        console.log("This test builds the image twice to demonstrate cache effectiveness\n");

        // First build (populates cache)
        console.log("Build 1: Populating sccache (slow)...");
        const start1 = Date.now();
        await DockerTestnet.buildLocalImage({
            cleanSccache: true, // Start fresh
        });
        const duration1 = (Date.now() - start1) / 1000;
        console.log(`✓ Build 1 completed in ${duration1.toFixed(1)}s\n`);

        // Second build (uses cache)
        console.log("Build 2: Using sccache (fast)...");
        const start2 = Date.now();
        await DockerTestnet.buildLocalImage({
            showStats: true, // Show cache statistics
        });
        const duration2 = (Date.now() - start2) / 1000;
        console.log(`✓ Build 2 completed in ${duration2.toFixed(1)}s\n`);

        // Compare
        const speedup = duration1 / duration2;
        console.log(`Speedup: ${speedup.toFixed(1)}x faster`);
        console.log(`Cache effectiveness: ${((1 - duration2/duration1) * 100).toFixed(1)}% time saved`);

        // Should be significantly faster (at least 2x)
        expect(speedup).toBeGreaterThan(2);
    }, 2400000); // 40 minute timeout (two full builds)
});

File: source/docker-testnet/rust-sdk/tests/local_build_test.rs
/// Local Image Build Integration Test
///
/// This test validates the entire local build workflow:
/// 1. Building the validator image from source
/// 2. Starting a testnet with the locally built image
/// 3. Verifying the testnet is fully functional
///
/// NOTE: This test is SLOW on first run (~10-15min to build image)
///       Subsequent runs are faster (~2-3min) thanks to sccache
///
/// To run:
///   cargo test --test local_build_test -- --nocapture --test-threads=1
///
/// To run with logging:
///   RUST_LOG=info cargo test --test local_build_test -- --nocapture --test-threads=1

use atomica_docker_testnet::{DockerTestnet, BuildOptions};
use serial_test::serial;

#[tokio::test]
#[serial]
async fn test_build_local_image() {
    let _ = tracing_subscriber::fmt().with_test_writer().try_init();

    println!("\n=== Building Local Validator Image ===");
    println!("This may take 10-15 minutes on first build...");
    println!("Subsequent builds will be much faster (~2-3min) thanks to sccache\n");

    match DockerTestnet::build_local_image(Some(BuildOptions {
        profile: Some("release".to_string()),
        show_stats: false, // Set to true if you want sccache stats
        ..Default::default()
    }))
    .await
    {
        Ok(()) => {
            println!("✓ Image built successfully");
        }
        Err(e) => {
            eprintln!("✗ Failed to build image: {}", e);
            panic!("Build failed: {}", e);
        }
    }
}

#[tokio::test]
#[serial]
async fn test_start_testnet_with_local_image() {
    let _ = tracing_subscriber::fmt().with_test_writer().try_init();

    println!("\n=== Starting Testnet with Local Image ===");

    // First ensure the image is built
    // (This will be fast if test_build_local_image already ran)
    if let Err(e) = DockerTestnet::build_local_image(None).await {
        eprintln!("Note: Skipping test - failed to build image: {}", e);
        return;
    }

    // Start testnet with local image
    match DockerTestnet::new(2, true).await {
        Ok(testnet) => {
            println!("✓ Testnet started with {} validators", testnet.num_validators());
            println!("  Using image: atomica-validator:local");

            // Verify testnet is functional
            println!("\n=== Verifying Testnet Functionality ===");

            // Wait a bit for consensus to stabilize
            tokio::time::sleep(tokio::time::Duration::from_secs(5)).await;

            // Try to wait for blocks
            match testnet.wait_for_blocks(1, 60).await {
                Ok(()) => println!("✓ Consensus is running - blocks being produced"),
                Err(e) => println!("Warning: Timeout waiting for blocks: {}", e),
            }

            // Query each validator
            for i in 0..testnet.num_validators() {
                match testnet.get_ledger_info(i).await {
                    Ok(info) => {
                        println!(
                            "✓ Validator {}: chain_id={}, version={}, height={}",
                            i, info.chain_id, info.ledger_version, info.block_height
                        );

                        assert_eq!(info.chain_id, 4);
                        assert_eq!(info.node_role, "validator");
                        assert!(info.ledger_version > 0);

                        // Check for git hash (build metadata)
                        if let Some(git_hash) = info.git_hash {
                            println!("  Build metadata: git_hash={}", git_hash);
                            assert!(git_hash.len() > 6); // Should be a commit hash
                        }
                    }
                    Err(e) => {
                        eprintln!("✗ Failed to query validator {}: {}", i, e);
                        panic!("Query failed");
                    }
                }
            }

            println!("\n✅ All tests passed - local image is fully functional!");
        }
        Err(e) => {
            eprintln!("✗ Failed to start testnet: {}", e);
            panic!("Testnet start failed: {}", e);
        }
    }
}

#[tokio::test]
#[serial]
#[ignore] // Ignored by default - run with --ignored to test sccache effectiveness
async fn test_sccache_speedup() {
    let _ = tracing_subscriber::fmt().with_test_writer().try_init();

    println!("\n=== Testing sccache Performance ===");
    println!("This test builds the image twice to demonstrate cache effectiveness\n");

    // First build (populates cache)
    println!("Build 1: Populating sccache (slow)...");
    let start1 = std::time::Instant::now();

    DockerTestnet::build_local_image(Some(BuildOptions {
        clean_sccache: true, // Start fresh
        ..Default::default()
    }))
    .await
    .expect("First build failed");

    let duration1 = start1.elapsed();
    println!("✓ Build 1 completed in {:.1}s\n", duration1.as_secs_f64());

    // Second build (uses cache)
    println!("Build 2: Using sccache (fast)...");
    let start2 = std::time::Instant::now();

    DockerTestnet::build_local_image(Some(BuildOptions {
        show_stats: true, // Show cache statistics
        ..Default::default()
    }))
    .await
    .expect("Second build failed");

    let duration2 = start2.elapsed();
    println!("✓ Build 2 completed in {:.1}s\n", duration2.as_secs_f64());

    // Compare
    let speedup = duration1.as_secs_f64() / duration2.as_secs_f64();
    let time_saved = (1.0 - duration2.as_secs_f64() / duration1.as_secs_f64()) * 100.0;

    println!("Speedup: {:.1}x faster", speedup);
    println!("Cache effectiveness: {:.1}% time saved", time_saved);

    // Should be significantly faster (at least 2x)
    assert!(
        speedup > 2.0,
        "Expected at least 2x speedup, got {:.1}x",
        speedup
    );
}
